packages/frontend/src/providers/ExplorerProvider.tsx

```diff
@@ -458,6 +485,40 @@ function reducer(
                 },
             };
         }
+        case ActionType.ADD_SORT_FIELD: {
+            return produce(state, (newState) => {
+                const sort =
+                    newState.unsavedChartVersion.metricQuery.sorts.find(
+                        (sf) => sf.fieldId === action.payload.fieldId,
+                    );
+
+                if (sort) {
+                    sort.descending = action.payload.descending;
+                } else {
+                    newState.unsavedChartVersion.metricQuery.sorts.push(
+                        action.payload,
+                    );
+                }
+            });
+        }
+        case ActionType.REMOVE_SORT_FIELD: {
+            return produce(state, (newState) => {
+                newState.unsavedChartVersion.metricQuery.sorts =
+                    newState.unsavedChartVersion.metricQuery.sorts.filter(
+                        (sf) => sf.fieldId !== action.payload,
+                    );
+            });
+        }
+        case ActionType.SWAP_SORT_FIELDS: {
+            return produce(state, (newState) => {
+                const sorts = newState.unsavedChartVersion.metricQuery.sorts;
+                const source = sorts[action.payload.sourceIndex];
+                const destination = sorts[action.payload.destinationIndex];
+
+                sorts[action.payload.sourceIndex] = destination;
+                sorts[action.payload.destinationIndex] = source;
+            });
+        }
```

Comment:
I added immer.js (battle tested at facebook) because of the nested updates that were quite complicated in this file. for those who are not familiar with it - it exposes `produce` function which takes a value and callback and whatever mutations happen in the callback is returned as a new immutable object.

Reply:
Do you recommend us using this in all our reducers ?

Reply:
yeah if @lightdash/engineering is ok with that.
right now it's incredibly hard to understand what this action changes in the end

<img width="767" alt="CleanShot 2022-09-02 at 17 42 52@2x" src="https://user-images.githubusercontent.com/962095/188159324-2d74d33a-73c7-411f-a8e6-3c00befd809e.png">